### PR TITLE
SYS-3338: Convert ValidatorAccountIds to Bounded Vector

### PR DESCRIPTION
**Pull request by thadouk**
This commit converts the ValidatorAccountIds storage item to a bounded Vector.
The vector is now limited to 256 elements. The necessary adjustments have been
made to the operations that interact with the storage, using the appropriate interfaces.
No storage migration is required for this conversion.
Tests have been updated accordingly to reflect the changes.

### CODE REVIEW

File: pallets/validators-manager/src/lib.rs [hunk 1]
```diff
@@ -192,6 +190,7 @@ pub mod pallet {
 
     #[pallet::storage]
     #[pallet::getter(fn get_vote)]
+    #[pallet::unbounded]
```
Comment (alexpArtos) [inline pallets/validators-manager/src/lib.rs:193]: why do we need this?

File: pallets/validators-manager/src/lib.rs [hunk 2]
```diff
@@ -257,8 +256,7 @@ pub mod pallet {
         ) -> DispatchResultWithPostInfo {
             ensure_root(origin)?;
             let validator_account_ids =
-                Self::validator_account_ids().or_else(|| Some(vec![])).expect("empty vec");
-            ensure!(validator_account_ids.len() > 0, Error::<T>::NoValidators);
+                Self::validator_account_ids().ok_or(Error::<T>::NoValidators)?;
```
Comment (alexpArtos) [inline pallets/validators-manager/src/lib.rs:259]: why did we remove the explicit check that the length is > 0? What exactly does `ok_or` do in this case?

File: pallets/validators-manager/src/tests/tests.rs [hunk 3]
```diff
@@ -588,7 +588,9 @@ mod add_validator {
                 let context = &AddValidatorContext::default();
 
                 set_session_keys(&context.collator);
-                <<ValidatorManager as Store>::ValidatorAccountIds>::append(&context.collator);
+                assert_ok!(<<ValidatorManager as Store>::ValidatorAccountIds>::try_append(
```
Comment (alexpArtos) [inline pallets/validators-manager/src/tests/tests.rs:591]: Shouldn't we add a test in here for when we have the maximum number of collators and try to add beyond it?

Comment (thadouk) [reply pallets/validators-manager/src/tests/tests.rs:591]: What is the behaviour the test is testing?
The BoundedVector implementation?

File: pallets/validators-manager/src/lib.rs (hunk 1 above)
Comment (thadouk) [reply pallets/validators-manager/src/lib.rs:193]:
```
/// The optional attribute `#[pallet::unbounded]` declares the storage as unbounded. When
/// implementating the storage info (when `#[pallet::generate_storage_info]` is specified on
/// the pallet struct placeholder), the size of the storage will be declared as unbounded. This
/// can be useful for storage which can never go into PoV (Proof of Validity).
```

**Review by ivan-cholakov — APPROVED** [review]

File: pallets/validators-manager/src/tests/tests.rs (hunk 3 above)
Comment (alexpArtos) [reply pallets/validators-manager/src/tests/tests.rs:591]: No, actually not. We are doing functional testing, or in other words, behaviour-testing.

We are testing `add_validator`. If you see the sequence of contexts in the code, it would be something like this:

`add_validator fails_when validator_list_is_full`

So, you see, from the functionality's point of view, we don't care if we have bounded-vec or not. We now have a functional difference to what we had before, (a limit on addition) but we don't have that in the tests. In this case, the correct behaviour would be to receive a MaxValidators-type error.

It's a different function contract, so I think we should have a test to represent that.

Comment (thadouk) [reply pallets/validators-manager/src/tests/tests.rs:591]: I see your point, although I think the test is of little value. We don't do such extensive testing on our endpoints.
The mock effort required trumps the value the test adds, as reaching this number of collator is academic.
The bound is not there to add such functionality to the system, but to align the implementation to the framework guidelines.
That being said I will add the test to accommodate the request as it probably won't be that complicated.
This won't be the case though for every bounded vector we convert.

File: pallets/validators-manager/src/lib.rs [hunk 4]
```diff
@@ -234,7 +233,7 @@ pub mod pallet {
                 self.validators.len()
             );
             for (validator_account_id, eth_public_key) in &self.validators {
-                <ValidatorAccountIds<T>>::append(validator_account_id);
+                assert_ok!(<ValidatorAccountIds<T>>::try_append(validator_account_id));
```
Comment (nahuseyoum) [inline pallets/validators-manager/src/lib.rs:236]:
```suggestion
                <ValidatorAccountIds<T>>::try_append(validator_account_id).expect("Too many validator accounts in genesis");
```

File: pallets/validators-manager/src/lib.rs [hunk 5]
```diff
@@ -66,13 +66,13 @@ use crate::offence::{
 #[frame_support::pallet]
 pub mod pallet {
     use super::*;
-    use frame_support::pallet_prelude::*;
+    use frame_support::{assert_ok, pallet_prelude::*};
```
Comment (nahuseyoum) [inline pallets/validators-manager/src/lib.rs:69]:
```suggestion
    use frame_support::pallet_prelude::*;
```

**Review by alexpArtos — APPROVED** [review]

File: pallets/validators-manager/src/lib.rs (hunk 1 above)
Comment (alexpArtos) [reply pallets/validators-manager/src/lib.rs:193]: I think applies in a Map, not a Vec. We are not making maps bounded.

Comment (nahuseyoum) [reply pallets/validators-manager/src/lib.rs:193]: If all we have is a map then we shouldn't use this attribute because the idea behind it is to explicitly mention that a storage has `unbounded` items and is usually used with storage items that store vectors.

Comment (thadouk) [reply pallets/validators-manager/src/lib.rs:193]: The storage item is unbounded cause the map value items, contain unbounded vectors. This is addressed in the voting changes in #178 
The attribute is optional, and it was added to be informative.
Nonetheless, I can see that since it creates confusion, I will remove it since it is removed anyway on #178

File: pallets/validators-manager/src/lib.rs (hunk 4 above)
Comment (thadouk) [reply pallets/validators-manager/src/lib.rs:236]: I prefer the assertion as it is more compact.

Comment (nahuseyoum) [reply pallets/validators-manager/src/lib.rs:235]: But the assertion doesn't give you a reason for the error. This one is more explicit and provides a meaningful error.

**Review by nahuseyoum — APPROVED** [review]
